text/0000-opaque-data.md

```diff
@@ -0,0 +1,514 @@
+struct, similarialy to other types in `core::marker` such as `PhantomData`.
+A crate defining FFI bindings to an opaque C type `Foo` would define the type
+as:
+
+```rust
+use std::marker; // or core::marker in #![no_std]
+
+#[repr(C)]
+pub struct Foo(marker::Opaque);
+```
+
+This type would have the following properties:
+
+1. The type would be `!Sized`.
+
+2. The type would be unconstructable, as it has a private constructor, (although
+   code inside `core::marker` should also not be able to construct a value of
+   the type).
+
+3. References and pointers to values of this type would be thin, having only a
+   single pointer width.
+
+4. A `Box<>` of this value would not free memory when dropped, like with ZSTs,
+   but would invoke the type's `drop(&mut self)` method if present.
+
+### Traits
+
+`Opaque` should implement the `Debug` trait, which should produce the string
+'Opaque'. It shouldn't implement any other traits from the standard library.
+
+## `size_of_val` and `align_of_val`
+
+Rust provides 2 intrinsics which are problematic to the implementation of
+`Opaque`, `size_of_val` and `align_of_val`. These two methods take an arbitrary
+unbounded reference, and return a `usize` representing the size of the value. We
+cannot implement these functions for `Opaque`, as it has a completely unknown
+layout, which in some cases may not be determinable even at runtime.
+
+There are multiple options for solving this problem, but this section will only
+cover one of them. The others are covered in the [Alternatives](#alternatives)
+section.
+
+## The `DynSized` trait
+
+A new built-in, unsafe, autoderived trait, `DynSized`, would be added. This type
+would be implemented on all existing types, except for `Opaque` and types which
+contain a `Opaque` object.
+
+Like `Sized`, the `DynSized` will be implied in all generic bounds, meaning that
+the generic declaration `<T>` would require the `DynSized` bound. Relaxing the
+`DynSized` bound would be done with `<T: ?DynSized>`.
+
+In addition, trait objects would be modified. Normal trait objects (`Trait`)
+would imply the `DynSized` bound, and contain the vtable entries required to
+implement the `size_of_val` and `align_of_val` intrinsics. Types which do not
+implement `DynSized` would not be able to be converted to a trait object of this
+kind. The syntax `Trait + ?DynSized` could also be written which would relax
+this `DynSized` requirement, and not include the vtable entries for
+`size_of_val` and `align_of_val`. `Sized` trait object should be able to
+implicitly coerce to `DynSized` trait objects by offseting their vtable pointer
+past the size and align vtable members.
+
+The `Sized` trait would be modified to inherit from the `DynSized` trait, meaning
+that `?DynSized` implies `?Sized`.
+
+Finally, many trait parmaeters would be updated in the standard library to relax
+their trait bounds from `?Sized` to `?DynSized` which should be a
+backwards-compatible change.
```

Comment:
(nit: parmaeters ↦ parameters)

`size_of_val` is used indirectly whenever the type needs to manage allocation itself, so types like `Box` and `Rc` need to remain `T: ?Sized`. 

Traits like `Borrow` and `AsRef` should be loosened to `?DynSized`. `UnsafeCell` should also be relaxed to `?DynSized` provided that it is safe to transmute an `&mut Opaque` to `&UnsafeCell<Opaque>`.

It may not be an easy change.